libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathChangesComputer.kt

```diff
@@ -5,184 +5,94 @@
 import com.intellij.openapi.util.io.FileUtil
-import org.jetbrains.kotlin.build.report.BuildReporter
-import org.jetbrains.kotlin.build.report.ICReporter
-import org.jetbrains.kotlin.build.report.metrics.*
-import org.jetbrains.kotlin.cli.common.ExitCode
 import org.jetbrains.kotlin.incremental.*
-import java.io.File
-import org.jetbrains.kotlin.gradle.incremental.ChangesCollectorResult.Success
-import org.jetbrains.kotlin.gradle.incremental.ChangesCollectorResult.Failure
 import org.jetbrains.kotlin.incremental.storage.FileToCanonicalPathConverter
 import java.util.*
+import kotlin.collections.LinkedHashMap
 
 /** Computes [ClasspathChanges] between two [ClasspathSnapshot]s .*/
 object ClasspathChangesComputer {
 
-    fun getChanges(current: ClasspathSnapshot, previous: ClasspathSnapshot): ClasspathChanges {
-        val changesCollector = ChangesCollector()
-        return when (collectClasspathChanges(current, previous, changesCollector)) {
-            Success -> {
-                val (lookupSymbols, fqNames, _) = changesCollector.getDirtyData(emptyList(), NoOpBuildReporter)
-                ClasspathChanges.Available(lookupSymbols.toList(), fqNames.toList())
-            }
-            is Failure -> ClasspathChanges.NotAvailable.UnableToCompute
-        }
-    }
-
-    private fun collectClasspathChanges(
-        current: ClasspathSnapshot,
-        previous: ClasspathSnapshot,
-        changesCollector: ChangesCollector
-    ): ChangesCollectorResult {
-        if (current.classpathEntrySnapshots.size != previous.classpathEntrySnapshots.size) {
-            return Failure.AddedRemovedClasspathEntries
-        }
+    fun compute(currentClasspathSnapshot: ClasspathSnapshot, previousClasspathSnapshot: ClasspathSnapshot): ClasspathChanges {
+        val currentClassSnapshots = currentClasspathSnapshot.getClassSnapshots()
+        val previousClassSnapshots = previousClasspathSnapshot.getClassSnapshots()
 
-        for (index in current.classpathEntrySnapshots.indices) {
-            val result = collectClasspathEntryChanges(
-                current.classpathEntrySnapshots[index],
-                previous.classpathEntrySnapshots[index],
-                changesCollector
-            )
-            if (result is Failure) {
-                return result
-            }
-        }
-        return Success
-    }
-
-    private fun collectClasspathEntryChanges(
-        current: ClasspathEntrySnapshot,
-        previous: ClasspathEntrySnapshot,
-        changesCollector: ChangesCollector
-    ): ChangesCollectorResult {
-        if (current.classSnapshots.size != previous.classSnapshots.size) {
-            return Failure.AddedRemovedClasses
-        }
+        val workingDir =
+            FileUtil.createTempDirectory(this::class.java.simpleName, "_WorkingDir_${UUID.randomUUID()}", /* deleteOnExit */ true)
+        val incrementalJvmCache = IncrementalJvmCache(workingDir, /* targetOutputDir */ null, FileToCanonicalPathConverter)
```

Comment:
For the next PR(s), we can also explore storing the previous compilation in `IncrementalJvmCache` so that we can avoid populating it with the previous snapshot.

Reply:
That's a good idea. It's a bit tricky since we want to compare the changed parts of the classpath only, but I've noted your suggestion to explore later once we have more insights on what parts of the task are taking time.